network/src/peer/fuzzing.rs

```diff
@@ -0,0 +1,150 @@
+/// doesn't panic or leak memory when reading, deserializing, and handling messages
+/// from remote peers.
+pub fn fuzz(data: &[u8]) {
+    // Use the basic single-threaded runtime, since our current tokio version has
+    // a chance to leak memory and/or thread handles when using the threaded
+    // runtime and sometimes blocks when trying to shutdown the runtime.
+    //
+    // https://github.com/tokio-rs/tokio/pull/2649
+    let mut rt = tokio::runtime::Builder::new()
+        .basic_scheduler()
+        .enable_all()
+        .build()
+        .unwrap();
+    let executor = rt.handle().clone();
+
+    // We want to choose a constant peer id for _our_ peer id, since we will
+    // generate unbounded metrics otherwise and OOM during fuzzing.
+    let peer_id = PeerId::ZERO;
+    // However, we want to choose a random _remote_ peer id to ensure we _don't_
+    // have metrics logging the remote peer id (which would eventually OOM in
+    // production for public-facing nodes).
+    let remote_peer_id = PeerId::random();
+
+    // Mock data
+    let network_context = NetworkContext::mock_with_peer_id(peer_id);
+    let socket = ReadOnlyTestSocketVec::new(data.to_vec());
+    let metadata = ConnectionMetadata::new(
+        remote_peer_id,
+        ConnectionId::from(123),
+        NetworkAddress::mock(),
+        ConnectionOrigin::Inbound,
+        MessagingProtocolVersion::V1,
+        SupportedProtocols::from(
+            [
+                ProtocolId::ConsensusRpc,
+                ProtocolId::ConsensusDirectSend,
+                ProtocolId::MempoolDirectSend,
+                ProtocolId::StateSynchronizerDirectSend,
+                ProtocolId::DiscoveryDirectSend,
+                ProtocolId::HealthCheckerRpc,
+            ]
+            .iter(),
+        ),
+    );
+    let connection = Connection { socket, metadata };
+
+    let (connection_notifs_tx, connection_notifs_rx) = channel::new_test(8);
+    let max_concurrent_reqs = 8;
+    let max_concurrent_notifs = 8;
+    let channel_size = 8;
+
+    // Spin up a new `Peer` actor through the `NetworkProvider` interface.
+    let (network_reqs_tx, network_notifs_rx) = NetworkProvider::start(
+        network_context,
+        executor,
+        connection,
+        connection_notifs_tx,
+        max_concurrent_reqs,
+        max_concurrent_notifs,
+        channel_size,
+        MAX_FRAME_SIZE,
+    );
+
+    rt.block_on(async move {
+        // Wait for "remote" to disconnect (we read all data and socket read
+        // returns EOF), we read a disconnect request, or we fail to deserialize
+        // something.
+        let connection_notifs = connection_notifs_rx.collect::<Vec<_>>().await;
+
+        // ACK the "remote" d/c and drop our handle to the Peer actor. Then wait
+        // for all network notifs to drain out and finish.
+        drop(network_reqs_tx);
+        let network_notifs = network_notifs_rx.collect::<Vec<_>>().await;
+    });
```

Comment:
so two things here:

* If more than 8 messages are sent to any channel, might it block the fuzzer?
* I wish we could create a fake channel that just does nothing on send, this would be useful in other places too : o

Reply:
also, why would the peer disconnect here? I feel like this might run forever

Reply:
> If more than 8 messages are sent to any channel, might it block the fuzzer?

nope! just imagine this is like node.js when we use basic scheduler. it's all single-threaded (all tasks are running on one thread) but tokio still multiplexes across them (when one task can't make progress, it'll switch to another pending task). so if this task is waiting on a network notif, tokio will switch to another pending task (e.g. the one reading network messages) to continue making progress.

Reply:
Ah, I missed that the last read will return `Poll::Ready(Ok(0))` which is interpreted as `EOF`.

And to your answer, my understanding here is that `collect()` here will continuously read the stream, potentially unblocking the sender. Isn't there a `drain()` function to avoid collecting :P?

Reply:
yeah `collect()` will keep polling the stream until it finishes. no `drain()` function though : (